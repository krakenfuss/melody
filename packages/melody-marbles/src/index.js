import { hot as hotObservable, cold as coldObservable } from 'jest-marbles';

/**
 * A decorator for the values, signalling that a hot observable should be used.
 */
class HotMarbleValues {
    constructor(values) {
        this.values = values;
    }
}
/**
 * Marks a marble diagram as a hot observable
 * @param {Object} values
 */
export const hot = values => new HotMarbleValues(values);
export const cold = values => values;

const mergeLinesWithParts = lines =>
    lines.reduce(
        (acc, line, index) => (index > 0 ? acc + `[[${index}]]` + line : line),
        ''
    );

/**
 * A tagged string literal helper which parses a simple domain specific language for describing
 * a set of Rx.js marble diagrams.
 * It returns an object with all described streams and subscriptions.
 *
 * ## Usage
 * ```
 * const { streamA, streaming, sub } = marbles`
 *     streamA   = ---a--b---d---| ${hot(values)}
 *     streaming = --b--da--e--c--  ${{ a: 1, b: 2, c: 3, d: 4, e: 5 }}
 *     sub       = -----^----
 * `;
 * ```
 *
 * In the above example, `streamA` would be a hot observable, `streaming` would be a cold observable
 * and `sub` would be a subscription.
 *
 * @param {String[]} inputMarbles The individual parts of the string literal
 * @param {Object[]} parts The value parts of the string literal
 * @returns {Object} An object containing all defined streams and subscriptions
 */
export const marbles = (inputMarbles, ...parts) => {
    const mergedMarbles = mergeLinesWithParts(inputMarbles);
    const regex = /^\s*([a-zA-Z0-9_]+)\s*=\s*([a-zA-Z-^!|#()]+)\s*(?:\[\[([0-9])+\]\])?\s*$/gm;
    const results = {};
    let m;

    while ((m = regex.exec(mergedMarbles)) !== null) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === regex.lastIndex) {
            regex.lastIndex++;
        }

        // The result can be accessed through the `m`-variable.
        const [_, name, marble, valueIndex] = m;
        if (valueIndex === undefined) {
            // handle subscription
            results[name] = marble;
        } else {
            // handle emitting stream
            const value = parts[valueIndex - 1];
            if (value instanceof HotMarbleValues) {
                results[name] = hotObservable(marble, value.values);
            } else {
                results[name] = coldObservable(marble, value);
            }
        }
    }

    return results;
};
