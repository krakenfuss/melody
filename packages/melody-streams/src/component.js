/**
 * Copyright 2018 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { enqueueComponent } from 'melody-idom';
import { shallowEqual } from './util/shallowEqual';
import { BehaviorSubject, Subject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';

function Component(element) {
    // part of the public API
    this.el = element;
    this.refs = {};
    this.props = new BehaviorSubject({});
    this.updates = new Subject();
    this.subscriptions = [];
    this.state = {};
}

Object.assign(Component.prototype, {
    apply(props) {
        this.props.next(props);
        if (this.subscriptions.length === 0) {
            const t = this.getTransform({
                dispatch(eventName, detail, options = {}) {
                    const event = new CustomEvent(eventName, {
                        ...options,
                        detail,
                    });
                    this.el.dispatchEvent(event);
                },
                props: this.props,
                updates: this.updates,
                subscribe: obs => this.subscriptions.push(obs.subscribe()),
            });
            const s = t
                .pipe(distinctUntilChanged(shallowEqual))
                .subscribe(state => {
                    this.state = state;
                    enqueueComponent(this);
                });
            this.subscriptions.push(s);
        }
    },

    notify() {
        this.updates.next();
    },

    componentWillUnmount() {
        this.props.complete();
        this.updates.complete();
        this.subscriptions.forEach(sub => sub.unsubscribe());
        this.subscriptions.length = 0;
    },

    render() {},
});

const createComponentConstructor = Parent => {
    function ChildComponent(el) {
        Parent.call(this, el);
    }
    ChildComponent.prototype = Object.create(Parent.prototype, {
        constructor: { value: ChildComponent },
    });
    return ChildComponent;
};

export const baseCreateComponent = (transform, templateFnOrObj) => {
    const template = templateFnOrObj.render
        ? props => templateFnOrObj.render(props)
        : templateFnOrObj;
    const ChildComponent = createComponentConstructor(Component);
    ChildComponent.prototype.displayName =
        transform.name || template.name || template.displayName || 'Unknown';
    ChildComponent.prototype.render = function() {
        return template(this.state);
    };
    ChildComponent.prototype.getTransform = transform;
    return ChildComponent;
};

export const createComponent = (...args) => {
    if (args.length >= baseCreateComponent.length) {
        return baseCreateComponent.apply(null, args);
    }
    return (...args2) => createComponent.apply(null, args.concat(args2));
};
