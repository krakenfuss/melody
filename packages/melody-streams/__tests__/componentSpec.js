/**
 * Copyright 2018 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { assert } from 'chai';

import { render, unmountComponentAtNode } from 'melody-component';
import {
    elementOpen,
    elementClose,
    text,
    component,
    patchOuter,
} from 'melody-idom';
import { createComponent, combine } from '../src';
// import { createComponent, useEffect, useEffectOnce, useState } from '../src';
import { flush } from './util/flush';
import {
    createSubjects,
    next,
    testWith,
    testWithArguments,
} from './util/testHelpers';
import { last } from 'rxjs/operators';

const template = {
    render(_context) {
        elementOpen('div', null, null);
        text(_context.value);
        elementClose('div');
    },
};

describe('component', () => {
    it('createComponent should support currying', () => {
        const template1 = {
            render(_context) {
                elementOpen('div');
                elementOpen('span');
                text(_context.value);
                elementClose('span');
                elementClose('div');
            },
        };
        const template2 = {
            render(_context) {
                elementOpen('div');
                elementOpen('div');
                text(_context.value);
                elementClose('div');
                elementClose('div');
            },
        };
        const createHashtagComponent = createComponent(({ props }) => props);
        const SpanHashtagComponent = createHashtagComponent(template1);
        const DivHashtagComponent = createHashtagComponent(template2);

        const root1 = document.createElement('div');
        render(root1, SpanHashtagComponent, { value: 'foo' });

        const root2 = document.createElement('div');
        render(root2, DivHashtagComponent, { value: 'foo' });
    });
    it('should render when props have changed', () => {
        const root = document.createElement('div');
        const [subj] = createSubjects(1);
        const MyComponent = createComponent(({ props, updates }) => {
            updates.subscribe(...next(subj));
            return props;
        }, template);
        // Snapshot should have 2 entries
        testWithArguments(subj, render, [
            [root, MyComponent, { value: 'foo' }],
            [root, MyComponent, { value: 'bar' }],
        ]);
    });
    it('should not rerender when props have not changed', () => {
        const root = document.createElement('div');
        const [subj] = createSubjects(1);
        const MyComponent = createComponent(({ props, updates }) => {
            updates.subscribe(...next(subj));
            return props;
        }, template);
        // Snapshot should have 1 entry
        testWithArguments(subj, render, [
            [root, MyComponent, { value: 'foo' }],
            [root, MyComponent, { value: 'foo' }],
        ]);
    });
    it("should not rerender children when props haven't changed", () => {
        const childTemplate = {
            render(_context) {
                elementOpen('div', null, null);
                text('Hello');
                elementClose('div');
            },
        };
        const [subj] = createSubjects(1);
        const MyComponent = createComponent(({ props, updates }) => {
            updates.subscribe(...next(subj));
            return props;
        }, childTemplate);

        const parentTemplate = {
            render(_context) {
                elementOpen('div', null, null);
                component(MyComponent, '4');
                elementClose('div');
            },
        };
        const MyParentComponent = createComponent(
            ({ props }) => props,
            parentTemplate
        );

        const root = document.createElement('div');
        // Snapshot should have 1 entry
        testWithArguments(subj, render, [
            [root, MyParentComponent],
            [root, MyParentComponent],
        ]);
    });
    it('should replace components', () => {
        const template = {
            render(_context) {
                elementOpen('div', 'test', null);
                text(_context.text);
                elementClose('div');
            },
        };
        const root = document.createElement('div');
        const MyComponent = createComponent(({ props }) => props, template);
        const MyOtherComponent = createComponent(
            ({ props }) => props,
            template
        );

        render(root, MyComponent, { text: 'hello' });
        assert.equal(root.outerHTML, '<div>hello</div>');

        render(root, MyOtherComponent, { text: 'test' });
        assert.equal(root.outerHTML, '<div>test</div>');
    });
    it('should unmount replaced components', () => {
        const template = {
            render(_context) {
                elementOpen('div', 'test', null);
                text(_context.text);
                elementClose('div');
            },
        };
        const root = document.createElement('div');
        const [subj] = createSubjects(1);
        const MyComponent = createComponent(({ props, updates }) => {
            updates.pipe(last()).subscribe(...next(subj));
            return props;
        }, template);
        const MyOtherComponent = createComponent(
            ({ props }) => props,
            template
        );

        render(root, MyComponent, { text: 'hello' });
        assert.equal(root.outerHTML, '<div>hello</div>');
        // Should have 1 entry
        testWithArguments(subj, render, [
            [root, MyOtherComponent, { text: 'test' }],
        ]);
        assert.equal(root.outerHTML, '<div>test</div>');
    });
});
