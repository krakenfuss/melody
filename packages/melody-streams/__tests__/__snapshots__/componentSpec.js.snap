// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component should not rerender children when props haven't changed 1`] = `"[null]"`;

exports[`component should not rerender when props have not changed 1`] = `"[null]"`;

exports[`component should render when props have changed 1`] = `"[null,null]"`;

exports[`component should unmount replaced components 1`] = `"[null]"`;
